### DISCUSSION TITLE
Open source 'next-test-utils'

### DISCUSSION THREAD

**Discussion by bobaaaaa** [vercel/next.js#13389] [Ideas]
Hey everyone. I wonder if someone else feels the need for the `next-test-utils` package. Like the nextjs team we also write lots of integration tests besides basic unit tests. I find it very hard writing the test setup to start nextjs in prod mode. Luckily the nextjs team has this internal lib for their own integration tests.

Testing use cases:

- community plugins
- custom servers
- e2e HTML rendering of pages

**Comment by armandabric** [comment]
I'm starting to work on e2e in cypress our SSG nextjs site, and it seems that for now we have a lot work on our side to be able to do it:
- providing a fake API in an independent process to mock our API data on server side
- controlling a nextjs server to force the regenation of the same url between to test to isolate them

If the `next-test-utils` contain anything that could help with that it will be a big win fot the community :+1: 

In a more general note, it could also be interesting to have a guide in the documentation that explain how to do e2e testing of a SSG application. The SSG part of the stack make the orchestration of test process more hard

**Comment by Vadorequest** [comment]
We've integrated automated E2E testing using Cypress and GitHub Actions on [NRN](https://github.com/UnlyEd/next-right-now).

You can see it working in our PR - [See E2E failure](https://github.com/UnlyEd/next-right-now/pull/68#issuecomment-636325492) and it provides with Artifacts (video/screenshots) for easier debug

Our GitHub Actions config can be found here:
https://github.com/UnlyEd/next-right-now/tree/v2-mst-aptd-gcms-lcz-sty/.github/workflows

@Spy-Seth Also, it uses different stages (production/staging) so the whole env is duplicated and E2E can be tested on a completely separated env (no impact on prod env). The current E2E implementation also tests the production, but that can be skipped if E2E mocking production would be complex, like it seems to be your case)

> This isn't really related to the use of a `next-test-utils`, but it may help those seeking E2E testing examples.

**Comment by bobaaaaa** [comment]
For my current (monorepo) project I created a standalone package `next-test-utils`. I copied&pasted some of the code https://github.com/vercel/next.js/blob/canary/test/lib/next-test-utils.js migrated to TypeScript + restructured + refactored it. The current API looks like:

```tsx
import { nextApp, cli, browser, utils } from 'next-test-utils';

// starting/stopping a next app via cli
cli.runNextDev(...);
cli.runNextStart(...);
cli.killApp(...);

// starting/stopping via next app (const app = Next({ dev: false }))
nextApp.start(nextJsApp);
nextApp.stop(nextJsApp);

// low-level fetch + cheerio wrapper
browser.get(...);
browser.html(...);
browser.html$(...);

utils.getFreePort();
```

If have also some more features in mind. Like adding puppeteer.

@timneutkens @Timer I'am open to share ❤️

**Reply by holgerw** [reply]
I was searching for your package on npm. It seems you did not publish it.

Would you consider sharing it? Is this still your approach for integration testing? Did you migrate to a different testing solution?

I am doing integration tests for API routes in Jest and would like to have routing covered by the tests (especially verify extraction of page params of dynamic API routes).

Could I help to make this package public?

**Comment by balazsorban44** [comment]
Related #10967

**Comment by aralroca** [comment]
I miss some tools to test all `rewrites` and `redirects` in an easy way.

Something like this will be very useful:

```js
const page = new PageTest(nextConfig)
page('/example/bah').toBeStatus(404)
page('/example/redirect').toBeStatus(308).to('/example-redirect')
page('/example/rewrites').toBeStatus(200).to('/another-destination')
```

I think this could even be done without compiling with webpack nor raising any server, just by analyzing the `pages` folder structure and reading the `next.config.js` configuration.
